### DISCUSSION TITLE
Obs Media Input Actions

### DISCUSSION THREAD

**Discussion by enrik90h** [Palakis/obs-websocket#1210] [Q&A]
Hello,

On OBS 30
I want to pause, play, next, previous, a VLC playlist in OBS, and I don't know how to do it since it says DEPRECATED in the notes.

**Comment by enrik90h** [comment]
I tried this:
MediaInputActionTriggered
{"sceneName":"Scene1","inputName":"List1","mediaAction":OBS_WEBSOCKET_MEDIA_INPUT_ACTION_PAUSE}

And results:
"Error: Your request is missing the `mediaAction` field."

**Comment by Namias** [comment]
There are updates?

**Comment by eedefeed** [comment]
Hi, I've been trying to find answers to this question too. It seems it's actually **not** deprecated, according to this comment in a bug report from from 2022:

> You may want to take a look at the freshly documented enum for OBSMediaInputAction (don't mind the deprecated warnings, it's not), because that OBS_WEBSOCKET_MEDIA_INPUT_ACTION_START doesn't exist. You have to use OBS_WEBSOCKET_MEDIA_INPUT_ACTION_RESTART after stopping a Media Input

Full context in the original thread:
https://github.com/obsproject/obs-websocket/issues/1046

The problem experienced by the bug reporter could be helpful to resolving @enrik90h's follow-up question.

**Comment by dovaldev** [comment]
I don't know if my answer will help anyone, but I have used this to replay a media source from an input source:


```
ws.call(requests.TriggerMediaInputAction(
    inputName='MediaSource',
    mediaAction='OBS_WEBSOCKET_MEDIA_INPUT_ACTION_RESTART'
    
))
```

**Reply by suagea** [reply]
Thank you. That's very helpful.
import websockets
import hashlib
import base64
import json
import uuid
from obs_logger import OBSLogger  # 引入日志模块


class OBSWebSocketClient:
  def __init__(self, url=None, password=None, log_to_console=False, log_to_file=True):
    """
    初始化 OBS WebSocket 客户端
    :param url: WebSocket 服务器地址(如"ws://127.0.0.1:4455")
    :param password: 连接密码
    :param log_to_console: 是否输出日志到控制台
    :param log_to_file: 是否输出日志到文件
    """
    self.url = url
    self.password = password or ""  # 确保password为字符串
    self.websocket = None
    self.rpc_version = 1
    self.logger = OBSLogger(
      console_log=log_to_console, file_log=log_to_file
    )

  async def connect(self):
    """连接到 OBS WebSocket 并完成身份验证"""
    self.logger.info("尝试连接到 OBS WebSocket...")
    self.websocket = await websockets.connect(self.url)

    # 接收 Hello 消息
    hello_message = await self._receive()
    self.logger.debug(f"接收到 Hello 消息: {hello_message}")

    # 从Hello中获取rpcVersion
    self.rpc_version = hello_message.get("d", {}).get("rpcVersion", 1)
    auth_data = hello_message.get("d", {}).get("authentication")

    # 发送 Identify 消息
    await self._identify(auth_data)
    self.logger.info("连接成功")

  async def _identify(self, auth_data):
    """发送 Identify 请求，根据需要进行认证"""
    identify_payload = {
      "op": 1,
      "d": {
        "rpcVersion": self.rpc_version
      }
    }

    # 如果服务器需要认证
    if auth_data:
      if not self.password:
        self.logger.error("OBS WebSocket需要认证，但未提供密码")
        raise Exception("OBS WebSocket需要认证，但未提供密码")
      challenge = auth_data["challenge"]
      salt = auth_data["salt"]

      # 生成 Base64 Secret
      base64_secret = base64.b64encode(
        hashlib.sha256((self.password + salt).encode()).digest()
      ).decode()

      # 生成最终认证字符串
      auth_string = base64.b64encode(
        hashlib.sha256((base64_secret + challenge).encode()).digest()
      ).decode()

      identify_payload["d"]["authentication"] = auth_string

    self.logger.info("发送 Identify 消息...")
    await self._send(identify_payload)

    # 接收服务器的Identified响应
    response = await self._receive()
    self.logger.debug(f"Identify响应: {response}")

    if response.get("op") != 2:  # op=2为Identified消息
      self.logger.error("OBS WebSocket认证/识别失败")
      raise Exception("OBS WebSocket认证/识别失败")

  async def _send(self, payload):
    """发送消息到 OBS WebSocket"""
    if not self.websocket:
      self.logger.error("WebSocket 尚未连接")
      raise Exception("WebSocket 尚未连接")
    self.logger.debug(f"发送消息: {payload}")
    await self.websocket.send(json.dumps(payload))

  async def _receive(self):
    """从 OBS WebSocket 接收消息"""
    if not self.websocket:
      self.logger.error("WebSocket 尚未连接")
      raise Exception("WebSocket 尚未连接")
    message = await self.websocket.recv()
    self.logger.debug(f"接收消息: {message}")
    return json.loads(message)

  async def send_request(self, request_type, request_data=None):
    """
    发送基础请求并返回响应结果。
    :param request_type: 请求类型，例如 "GetCurrentProgramScene"。
    :param request_data: 可选的请求参数，默认为 None。
    :return: 响应数据。
    """
    request_id = str(uuid.uuid4())  # 使用uuid确保唯一性
    payload = {
      "op": 6,
      "d": {
        "requestType": request_type,
        "requestId": request_id,
        "requestData": request_data or {}
      }
    }
    await self._send(payload)

    # 等待响应
    while True:
      response = await self._receive()
      # 确认是请求响应消息（op=7）并匹配requestId
      if response.get("op") == 7 and response["d"]["requestId"] == request_id:
        self.logger.info(f"请求 '{request_type}' 的响应: {response}")
        return response["d"]

  async def close(self):
    """关闭 WebSocket 连接"""
    if self.websocket:
      self.logger.info("关闭 WebSocket 连接")
      await self.websocket.close()
      self.websocket = None

  # 高层功能抽象
  async def get_scene_list(self):
    """获取场景列表"""
    self.logger.info("请求获取场景列表...")
    response = await self.send_request("GetSceneList")
    scenes = response.get("responseData", {}).get("scenes", [])
    self.logger.info(f"场景列表: {[scene['sceneName'] for scene in scenes]}")
    return scenes

  async def set_current_scene(self, scene_name):
    """
    切换到指定的场景。
    :param scene_name: 目标场景名称
    """
    self.logger.info(f"切换到场景: {scene_name}")
    response = await self.send_request("SetCurrentProgramScene", {"sceneName": scene_name})
    if response.get("requestStatus", {}).get("result", False):
      self.logger.info(f"成功切换到场景: {scene_name}")
    else:
      comment = response.get('requestStatus', {}).get('comment', '未知错误')
      self.logger.error(f"切换场景失败: {comment}")
      raise Exception(f"切换场景失败: {comment}")

  async def get_scene_sources(self, scene_name):
    """
    获取指定场景的来源信息。
    :param scene_name: 场景名称
    :return: 场景来源信息列表
    """
    self.logger.info(f"请求获取场景 '{scene_name}' 的来源信息...")
    response = await self.send_request("GetSceneItemList", {"sceneName": scene_name})
    sources = response.get("responseData", {}).get("sceneItems", [])
    self.logger.info(f"场景 '{scene_name}' 的来源信息: {sources}")
    return sources

  async def control_media(self, input_name, action):
    """
    控制媒体输入源的播放状态。
    :param input_name: 媒体输入源名称
    :param action: 动作类型，可选值为 'OBS_WEBSOCKET_MEDIA_INPUT_ACTION_PLAY',
                   'OBS_WEBSOCKET_MEDIA_INPUT_ACTION_PAUSE',
                   'OBS_WEBSOCKET_MEDIA_INPUT_ACTION_STOP',
                   'OBS_WEBSOCKET_MEDIA_INPUT_ACTION_RESTART',
                   'OBS_WEBSOCKET_MEDIA_INPUT_ACTION_NEXT',
                   'OBS_WEBSOCKET_MEDIA_INPUT_ACTION_PREVIOUS'
    :return: None
    """
    self.logger.info(f"控制媒体输入源 '{input_name}' 执行动作 '{action}'")
    await self.send_request("TriggerMediaInputAction", {
      "inputName": input_name,
      "mediaAction": action
    })
